### PR TITLE
dispatcher: add simple unit tests

### PR DESCRIPTION
**Pull request by mrice32**
*Description*:
The dispatcher has no real unit testing of documented functionality.  As components that are linked into Envoy start to take advantage of the dispatcher, this functionality should have unit tests that clearly show that the methods work as desired.

*Risk Level*: Low

*Testing*: Added new tests.

### CODE REVIEW

File: test/common/event/dispatcher_impl_test.cc [hunk 1]
```diff
@@ -51,5 +53,77 @@ TEST(DispatcherImplTest, DeferredDelete) {
...
+
+  ~DispatcherImplTest() override {
+    dispatcher_->exit();
+    dispatcher_thread_->join();
+  }
+
+  std::unique_ptr<Thread::Thread> dispatcher_thread_;
+  DispatcherPtr dispatcher_;
+  std::mutex mu_;
+  int work_id_;
```
Comment (mattklein123) [inline test/common/event/dispatcher_impl_test.cc]: nit: prefer uint64_t here. In general I'm not sure what work_id really buys us. It seems like you could be using condition variables and atomic variables to check? (See comments below)

File: test/common/event/dispatcher_impl_test.cc [hunk 2]
```diff
@@ -51,5 +53,77 @@ TEST(DispatcherImplTest, DeferredDelete) {
...
+};
+
+TEST_F(DispatcherImplTest, Post) {
+  dispatcher_->post([this]() {
+    std::lock_guard<std::mutex> lock(mu_);
+    work_id_ = 1;
+  });
+
+  bool reached_value = false;
+  while (!reached_value) {
```
Comment (mattklein123) [inline test/common/event/dispatcher_impl_test.cc]: nit: This is basically going to spin until the post happens. Would it be better to use a condition variable for these types of tests?

File: test/common/event/dispatcher_impl_test.cc [hunk 3]
```diff
@@ -51,5 +53,77 @@ TEST(DispatcherImplTest, DeferredDelete) {
...
+    std::lock_guard<std::mutex> lock(mu_);
+    timer = dispatcher_->createTimer([this]() {
+      std::lock_guard<std::mutex> lock(mu_);
+      work_id_ = 1;
+    });
+
+  });
+
+  bool reached_value = false;
+  while (!reached_value) {
```
Comment (mattklein123) [inline test/common/event/dispatcher_impl_test.cc]: Same spinning comment as above.

**Comment by mrice32** [comment]
I chose not to use atomics because of the issue mentioned [here](https://stackoverflow.com/a/36130475) and the fact that the condition variable requires a mutex to exist anyway.  Let me know if I'm missing something and using an atomic bool would be a better solution.

**Comment by mattklein123** [comment]
> I chose not to use atomics because of the issue mentioned here and the fact that the condition variable requires a mutex to exist anyway. Let me know if I'm missing something and using an atomic bool would be a better solution.

That's fine.

File: test/common/event/dispatcher_impl_test.cc [hunk 4]
```diff
@@ -51,5 +54,71 @@ TEST(DispatcherImplTest, DeferredDelete) {
...
+    {
+      std::lock_guard<std::mutex> lock(mu_);
+      work_finished_ = true;
+    }
+    cv_.notify_one();
+  });
+
+  std::unique_lock<std::mutex> lock(mu_);
+
+  EXPECT_TRUE(cv_.wait_for(lock, std::chrono::seconds(5), [this]() { return work_finished_; }));
```
Comment (mattklein123) [inline test/common/event/dispatcher_impl_test.cc]: CI can get absurdly slow and I could see this flaking sadly. My preference would be to not have a timeout and just let it hang if there is an issue, or I would use a much longer sanity timeout like 60s or something. Same below.

File: test/common/event/dispatcher_impl_test.cc [hunk 5]
```diff
@@ -51,5 +54,71 @@ TEST(DispatcherImplTest, DeferredDelete) {
...
+        }
+        cv_.notify_one();
+      });
+    }
+    cv_.notify_one();
+  });
+
+  std::unique_lock<std::mutex> lock(mu_);
+  ASSERT_TRUE(
+      cv_.wait_for(lock, std::chrono::seconds(5), [&timer]() { return static_cast<bool>(timer); }));
```
Comment (mattklein123) [inline test/common/event/dispatcher_impl_test.cc]: nit: If you do timer != nullptr you can lose the static case.

**Pull request merged**